cleanlab/experimental/datalab/issue_manager/duplicate.py

```diff
@@ -0,0 +1,147 @@
+
+    issue_name: str = "near_duplicate"
+
+    def __init__(
+        self,
+        datalab: Datalab,
+        metric: Optional[str] = "cosine",
+        threshold: Optional[float] = None,
+        k: Optional[int] = 10,
+        **_,
+    ):
+        super().__init__(datalab)
+        self.metric = metric
+        self.threshold = threshold
+        self.k = k
+        self.knn: Optional[NearestNeighbors] = None
+        self.near_duplicate_sets: List[List[int]] = []
+
+    def find_issues(
+        self,
+        features: List[str],
+        **_,
+    ) -> None:
+
+        feature_array = self._extract_embeddings(features)
+        if self.knn is None:
+            self.knn = NearestNeighbors(n_neighbors=self.k, metric=self.metric)
+
+        try:
+            check_is_fitted(self.knn)
+        except:
+            self.knn.fit(feature_array)
+
+        scores, distances = self._score_features(feature_array)
+        self.radius, self.threshold = self._compute_threshold_and_radius()
+
+        self.issues = pd.DataFrame(
+            {
+                f"is_{self.issue_name}_issue": scores < self.threshold,
+                self.issue_score_key: scores,
+            },
+        )
+
+        indices = self.knn.radius_neighbors(feature_array, self.radius, return_distance=False)
+        self.near_duplicate_sets = [
+            duplicates[duplicates != idx] for idx, duplicates in enumerate(indices)
+        ]
+        self.distances = distances
+
+        self.summary = self.get_summary(score=scores.mean())
+        self.info = self.collect_info()
+
+    def collect_info(self) -> dict:
+        issues_dict = {
+            "num_near_duplicate_issues": len(self.near_duplicate_sets),
+            "average_near_duplicate_score": self.issues[self.issue_score_key].mean(),
+            "near_duplicate_sets": self.near_duplicate_sets,
+            "radius": self.radius,
+        }
+
+        params_dict = {
+            "metric": self.metric,
+            "k": self.k,
+            "threshold": self.threshold,
+        }
+
+        knn = self.knn
+        dists, nn_ids = [array[:, 0] for array in knn.kneighbors()]  # type: ignore[union-attr]
+        weighted_knn_graph = knn.kneighbors_graph(mode="distance").toarray()  # type: ignore[union-attr]
+
+        # TODO: Reverse the order of the calls to knn.kneighbors() and knn.kneighbors_graph()
+        #   to avoid computing the (distance, id) pairs twice.
+        knn_info_dict = {
+            "nearest_neighbour": nn_ids.tolist(),
+            "distance_to_nearest_neighbour": dists.tolist(),
+            # TODO Check scipy-dependency
+            "weighted_knn_graph": weighted_knn_graph.tolist(),
+        }
+
+        info_dict = {
+            **issues_dict,
+            **params_dict,
+            **knn_info_dict,
+        }
+        return info_dict
+
+    def _extract_embeddings(self, columns: Union[str, List[str]], **kwargs) -> np.ndarray:
```

Comment:
If I remember right, this function is copied in multiple issue managers. Does it make sense to place it somewhere else for all issue managers to inherit?

Reply:
I've extracted these methods to the `datalab.data` module.
https://github.com/cleanlab/cleanlab/blob/d1d5edf56ab5a04ff6681d10cc553e5922e8c035/cleanlab/experimental/datalab/data.py#L89